### PR TITLE
Evaluate computed class props only once

### LINKED ISSUE #6240: Computed class fields evaluated during construction
> Choose one: is this a bug report or feature request?

bug report

### Input Code

```js
var x = 'x';

class F {
  [x] = 1;
  constructor() { 
    console.log(this.x);
  }
}

new F;
x = 'y';
new F;
```

### Babel Configuration (.babelrc, package.json, cli command)

stage3, stage2 presets

### Expected Behavior
1
1

### Current Behavior
1
undefined

### Your Environment
[REPL](https://babeljs.io/repl/build/4898/#?babili=false&browsers=&build=&builtIns=false&code_lz=G4QwTgBAHhC8EHIoINwCg0GMA2IDOeEAYhAN5oQQDaUAunBAIzqWYD2AdngC5gCumbmzAAKAJRkIFShHZc22AKYA6bGwDmI7gAsAlnmVQxLCAF805tB0UB3YuhjwEAT1RXb9oA&debug=false&circleciRepo=&evaluate=true&lineWrap=false&presets=stage-2%2Cstage-3&prettier=false&targets=&version=7.0.0-beta.0)

### PR DESCRIPTION
**Pull request by Qantas94Heavy**
| Q                        | A
| ------------------------ | ---
| Fixed Issues?            | Fixes #6240
| Patch: Bug Fix?          | yes
| Major: Breaking Change?  | no
| Minor: New Feature?      | no
| Tests Added + Pass?      | Yes
| Documentation PR         | no
| Any Dependency Changes?  | no

Previously, computed class properties would be evaluated every time a
new instance of the class was created. This means the property name
may have changed between different instances, as well as potential side
effects.

This commit fixes this by storing the computed value in a separate variable.

### CODE REVIEW

File: packages/babel-plugin-transform-class-properties/src/index.js [hunk 1]
```diff
@@ -96,6 +96,18 @@ export default function({ types: t }, options) {
           if (isStatic) {
             nodes.push(buildClassProperty(ref, propNode, path.scope));
           } else {
+            // Make sure computed property names are only evaluated once (upon
+            // class definition).
+            if (propNode.computed) {
+              const ident = path.scope.generateUidIdentifier("computed");
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-class-properties/src/index.js]: Here you could use `path.scope.generateUidIdentifierBasedOnNode(propNode.value, "computed")`, which generates better names.

e.g. `class A { [x.y] = 2; }` should become something like:
```js
var A = function A() {
  _classCallCheck(this, A);

  Object.defineProperty(this, _x$y, {
    configurable: true,
    enumerable: true,
    writable: true,
    value: 2
  });
};

var _x$y = x;
```

Comment (Qantas94Heavy) [reply packages/babel-plugin-transform-class-properties/src/index.js]: Ah, didn't know about that. Thanks!

**Comment by nicolo-ribaudo** [comment]
Thank you!

File: packages/babel-plugin-transform-class-properties/src/index.js [hunk 2]
```diff
@@ -96,6 +96,20 @@ export default function({ types: t }, options) {
...
+            if (propNode.computed) {
+              const ident = path.scope.generateUidIdentifierBasedOnNode(
+                propNode.key,
+              );
+              nodes.push(
+                t.variableDeclaration("var", [
+                  t.variableDeclarator(ident, propNode.key),
+                ]),
+              );
+              propNode.key = ident;
```
Comment (jridgewell) [inline packages/babel-plugin-transform-class-properties/src/index.js]: This needs to be cloned.

**Pull request merged**